File: 程式設計：資料結構答案檔/C#/304.cs
using System;
using System.Collections.Generic;

class Program
{
    static void Main()
    {
        int m = int.Parse(Console.ReadLine());
        List<string> fabrics = new List<string>();

        for (int i = 0; i < m; i++)
        {
            fabrics.Add(Console.ReadLine());
        }

        fabrics.Sort();

        int ans = 0;
        foreach (var s1 in fabrics)
        {
            int len = s1.Length;
            for (int i = 1; i <= len / 2; i++)
            {
                if (s1.Substring(0, i) == s1.Substring(len - i, i))
                {
                    string subStr = s1.Substring(i, len - 2 * i);
                    if (fabrics.BinarySearch(subStr) >= 0)
                    {
                        ans++;
                    }
                }
            }
        }

        Console.WriteLine(ans);
    }
}

File: 程式設計：資料結構答案檔/C++/304.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int main(void)
{
	int m;
	string s0;
	vector<string> v;
	int len;
	int i;
	int ans = 0;
	
	cin >> m;
	while(m--){
		cin >> s0;
		v.push_back(s0);
	} 
	sort(v.begin(), v.end());
	
    for(auto s1:v){
    	len = s1.length(); 
    	for(i=1; i<=len/2; i++){
    		if(s1.substr(0, i) == s1.substr(len-i, i)){
				if(binary_search(v.begin(), v.end(), 
				                 s1.substr(i, len-2*i))){
					ans++;
				}
			}
		}
	}
	cout << ans; 

	return 0;
}